test/jdk/java/awt/im/4959409/bug4959409.java

```diff
@@ -22,51 +22,180 @@
+            frame.setVisible(true);
+        });
+
+        boolean isFrameVisible = frameVisibleLatch.await(TIMEOUT, TimeUnit.SECONDS);
+        if (!isFrameVisible) {
+            throw new RuntimeException("Frame is not visible after " + TIMEOUT + "  seconds");
+        }
+
+        if (!isComponentVisible(jTextField, componentVisibleLatch)) {
+            throw new RuntimeException("Component is not visible after " + TIMEOUT + "   seconds");
+        }
+
+        Robot robot = new Robot();
+        robot.delay(DELAY);
+        robot.waitForIdle();
+
+        SwingUtilities.invokeAndWait(() -> {
+            points[0] = jTextField.getLocationOnScreen();
+            rect[0] = jTextField.getBounds();
+        });
+
+        performMouseAction(robot, points[0].x + rect[0].width / 2, points[0].y + rect[0].height / 2);
+
+        // Press SHIFT + 1 keys
+        robot.delay(DELAY);
+        robot.keyPress(KeyEvent.VK_SHIFT);
+        robot.delay(DELAY);
+        robot.keyPress(KeyEvent.VK_1);
+        robot.delay(DELAY);
+        robot.keyRelease(KeyEvent.VK_SHIFT);
+        robot.delay(DELAY);
+        robot.keyRelease(KeyEvent.VK_1);
+        robot.delay(DELAY);
+        robot.waitForIdle();
+
+        if (!keyPressedEventLatch.await(TIMEOUT, TimeUnit.SECONDS)) {
+            throw new RuntimeException("Did not received KEY PRESS for Shift + 1 , test failed");
+        }
+        return frame;
     }
-}
 
-class TestFrame extends JFrame implements KeyListener {
-    JTextField text;
-    JLabel label;
-
-    TestFrame () {
-        text = new JTextField();
-        text.addKeyListener(this);
-        label = new JLabel(" ");
-        Container c = getContentPane();
-        BorderLayout borderLayout1 = new BorderLayout();
-        c.setLayout(borderLayout1);
-        c.add(text, BorderLayout.CENTER);
-        c.add(label, BorderLayout.SOUTH);
-        setSize(300, 200);
-        setVisible(true);
+    public static void performMouseAction(final Robot robot, final int X, final int Y) {
+        robot.waitForIdle();
+        robot.delay(DELAY);
+        robot.mouseMove(X, Y);
+        robot.delay(DELAY);
+        robot.waitForIdle();
+
+        robot.delay(DELAY);
+        robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+        robot.delay(DELAY);
+        robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+        robot.delay(DELAY);
+        robot.waitForIdle();
     }
 
-    public void keyPressed(KeyEvent e) {
-        int code = e.getKeyCode();
-        int mods = e.getModifiers();
-        if (code == '1' && mods == KeyEvent.SHIFT_MASK) {
-            label.setText("KEYPRESS received for Shift+1");
+    public static boolean isComponentVisible(java.awt.Component component, CountDownLatch componentVisibleLatch) throws Exception {
+        int count = 1;
+        if (!component.isVisible()) {
+            while (count <= 5) {
+                TimeUnit.SECONDS.sleep(1);
+                if (component.isVisible()) {
```

Comment:
Should `component.isVisible()` also be called on EDT?

Reply:
done